compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -1,38 +1,412 @@
+    fn get_operators_iter(&self) -> impl Iterator<Item = VariancedDagOperator<'_>> {
+        self.dag.get_indices_iter().map(|i| self.get_operator(i))
+    }
+
+    fn get_operators_iter_mut(&mut self) -> impl Iterator<Item = VariancedDagOperatorMut<'_>> {
+        struct Iter<'a> {
+            dag: &'a mut VariancedDag,
+            range: Range<usize>,
+        }
+
+        impl<'a> Iterator for Iter<'a> {
+            type Item = VariancedDagOperatorMut<'a>;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                let Range { start, end } = self.range;
+                if start == end {
+                    return None;
+                }
+                self.range = (start + 1)..end;
+                unsafe {
+                    Some(std::mem::transmute::<
+                        VariancedDagOperatorMut<'_>,
+                        VariancedDagOperatorMut<'a>,
+                    >(
+                        self.dag.get_operator_mut(OperatorIndex(start))
+                    ))
+                }
+            }
+        }
+        Iter {
+            range: 0..self.dag.len(),
+            dag: self,
+        }
+    }
+
+    #[allow(unused)]
+    fn get_composed_inputs_iter(&self) -> impl Iterator<Item = ComposedInput<'_>> {
+        self.dag
+            .composition
+            .clone()
+            .into_iter()
+            .map(move |(id, from)| ComposedInput {
+                dag: self,
+                id,
+                from,
+            })
+    }
+
+    fn get_composed_inputs_iter_mut(&mut self) -> impl Iterator<Item = ComposedInputMut<'_>> {
+        struct Iter<'a> {
+            dag: &'a mut VariancedDag,
+            iter: std::collections::hash_map::IntoIter<OperatorIndex, Vec<OperatorIndex>>,
+        }
+
+        impl<'a> Iterator for Iter<'a> {
+            type Item = ComposedInputMut<'a>;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                self.iter.next().map(|(to, from)| unsafe {
+                    std::mem::transmute::<ComposedInputMut<'_>, ComposedInputMut<'a>>(
+                        ComposedInputMut {
+                            dag: self.dag,
+                            id: to,
+                            from,
+                        },
+                    )
+                })
+            }
+        }
+
+        Iter {
+            iter: self.dag.composition.clone().into_iter(),
+            dag: self,
+        }
+    }
+
+    fn get_operator(&self, index: OperatorIndex) -> VariancedDagOperator<'_> {
+        VariancedDagOperator {
+            dag: self,
+            id: index,
+        }
+    }
+
+    fn get_operator_mut(&mut self, index: OperatorIndex) -> VariancedDagOperatorMut<'_> {
+        VariancedDagOperatorMut {
+            dag: self,
+            id: index,
+        }
+    }
+
+    /// Patches the inputs following the composition rules.
+    fn apply_composition_rules(&mut self) {
+        for compo in self.get_composed_inputs_iter_mut() {
```

Comment:
A lot of complicated code just to be able to loop and apply composition rules.
ComposedInput + ComposedInputMut + get_composed_inputs_iter + get_composed_inputs_iter_mut
If you inline here without all the intermediate types it's probably simpler to follow and with 4x less lines.

Reply:
I agree that this is not the best part of the PR, and that there may be a better way to do that, but I find your comment a bit too harsh for my taste :) . My perspective is that currently, there is a huge lot of implicit knowledge needed to understand the code of the optimizer. Putting some knowledge in the type system via types, iterators, and whatnot is the preferred solution when using the rust language. Also simpler depends a lot about your habits. If you use rust analyzer it is very easy to navigate a codebase with a lot of types and iterators. Lastly, I don't think we should strive for a 4x condensed codebase, if it means being implicit and cryptic.